dask/dataframe/multi.py

```diff
@@ -230,7 +230,51 @@ def require(divisions, parts, required=None):
 
 def merge_chunk(lhs, *args, **kwargs):
     empty_index_dtype = kwargs.pop("empty_index_dtype", None)
-    out = lhs.merge(*args, **kwargs)
+
+    rhs = args[0] if len(args) > 0 else None
+    left_index = kwargs.get("left_index", False)
+    right_index = kwargs.get("right_index", False)
+
+    # discover categoricals
+    left_cats = lhs.select_dtypes(include="category")
+    right_cats = rhs.select_dtypes(include="category")
+
+    for col in set(left_cats.columns).union(right_cats.columns):
+        left = None
+        right = None
+
+        if col in left_cats:
+            left = left_cats[col]
+        elif col == kwargs.get("right_on", None) and left_index:
+            left = lhs.index
+
+        if col in right_cats:
+            right = right_cats[col]
+        elif col == kwargs.get("left_on", None) and right_index:
+            right = rhs.index
+
+        dtype = "category"
+        if left is not None and right is not None:
+            dtype = union_categoricals(
+                [left.astype("category"), right.astype("category")]
+            )
+
+        if left is not None:
+            if isinstance(left, pd.Index):
+                lhs.index = left.astype(dtype)
+            else:
+                lhs[col] = left.astype(dtype)
+        if right is not None:
+            if isinstance(right, pd.Index):
+                rhs.index = right.astype(dtype)
+            else:
+                rhs[col] = right.astype(dtype)
+
+    if rhs is not None:
+        out = lhs.merge(rhs, *args[1:], **kwargs)
+    else:
+        out = lhs.merge(*args, **kwargs)
```

Comment:
Can maybe be simplified to

```suggestion
    out = lhs.merge(rhs, *args, **kwargs)
```